File: Practica 3/Cola.py
class Cola:
    def __init__(self, n):
        self.n = n
        self.arreglo = [0] * n
        self.inicio = 0
        self.fin = -1
        self.elementos = 0

    def insert(self, e):
        if not self.isFull():
            self.fin = (self.fin + 1) % self.n
            self.arreglo[self.fin] = e
            self.elementos += 1
        else:
            print("Cola llena")

    def remove(self):
        if not self.isEmpty():
            valor = self.arreglo[self.inicio]
            self.inicio = (self.inicio + 1) % self.n
            self.elementos -= 1
            return valor
        print("Cola vacía")
        return -1

    def peek(self):
        if not self.isEmpty():
            return self.arreglo[self.inicio]
        print("Cola vacía")
        return -1

    def isEmpty(self):
        return self.elementos == 0

    def isFull(self):
        return self.elementos == self.n

    def size(self):
        return self.elementos


# Prueba de la clase con 3 elementos
if __name__ == "__main__":
    cola = Cola(5)
    cola.insert(10)
    cola.insert(20)
    cola.insert(30)
    #Agrega un elemento al final de la cola.
    
    print(cola.remove())  # 10 Elimina y devuelve el primer elemento.
    print(cola.remove())  # 20
    print(cola.peek())    # 30 Muestra el primer elemento sin eliminarlo.

File: Practica 3/Cola/src/cola.java
class cola {
    private long[] arreglo;
    private int inicio, fin, n, elementos;

    public cola(int n) {
        this.n = n;
        this.arreglo = new long[n];
        this.inicio = 0;
        this.fin = -1;
        this.elementos = 0;
    }

    public void insert(long e) {
        if (!isFull()) {
            fin = (fin + 1) % n;
            arreglo[fin] = e;
            elementos++;
        } else {
            System.out.println("Cola llena");
        }
    }

    public long remove() {
        if (!isEmpty()) {
            long valor = arreglo[inicio];
            inicio = (inicio + 1) % n;
            elementos--;
            return valor;
        }
        System.out.println("Cola vacía");
        return -1;
    }

    public long peek() {
        if (!isEmpty()) {
            return arreglo[inicio];
        }
        System.out.println("Cola vacía");
        return -1;
    }

    public boolean isEmpty() {
        return elementos == 0;
    }

    public boolean isFull() {
        return elementos == n;
    }

    public int size() {
        return elementos;
    }

    public static void main(String[] args) {
        cola cola = new cola(5);
        cola.insert(10);
        cola.insert(20);
        cola.insert(30);
        System.out.println(cola.remove()); // 10
        System.out.println(cola.remove()); // 20
        System.out.println(cola.peek());   // 30
    }
}
